subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -153,6 +153,16 @@ private SnapshotHierarchy removeUnwatchedContentAtNewWatchableHierarchy(String u
         return newRoot;
     }
 
+    /**
+     * Removes all content which is not watched under an existing snapshot hierarchy.
+     *
+     * When Gradle starts watching a hierarchy, it may be that there is already some existing content in the VFS.
+     * For example this happens when Gradle references a file in an included build directly by path,
+     * even before the included build is declared.
```

Comment:
This sounds like a fishy thing to do. What is a legitimate example of doing this? Can we warn/deprecate/fail instead?

Reply:
The use-case Cedric found in the original issue (https://github.com/gradle/gradle/issues/17621): The `java-gradle-plugin` adds the jar file of the composite build to the build script classpath in the test. Then the test itself also includes the build which contains the JAR file.

The use-case is somewhat exotic, though I can imagine there are a few more exotic cases which may cause this, especially when a test which includes a plugin from an included build is involved.